modules/adspiritBidAdapter.js

```diff
@@ -0,0 +1,129 @@
+ {
+  let requests = [];
+  let bidRequest;
+  let reqUrl;
+  let placementId;
+  for(let i = 0; i < validBidRequests.length; i++)
+  {
+   bidRequest = validBidRequests[i];
+   bidRequest.adspiritConId = spec.genAdConId(bidRequest);
+   reqUrl = spec.getBidderHost(bidRequest);
+   placementId = utils.getBidIdParameter('placementId', bidRequest.params);
+   reqUrl = '//' + reqUrl + RTB_URL + '&pid=' + placementId + '&ref=' + encodeURIComponent(bidderRequest.refererInfo.topmostLocation) + '&scx=' + (screen.width) + '&scy=' + (screen.height) + '&wcx=' + ('innerWidth' in window ? window.innerWidth : page.clientWidth) + '&wcy=' + ('innerHeight' in window ? window.innerHeight : page.clientHeight) + '&async=' + bidRequest.adspiritConId + '&t=' + Math.round(
+    Math.random() * 100000);
+   requests.push({
+                  method    : 'GET',
+                  url       : reqUrl,
+                  data      : {},
+                  bidRequest: bidRequest
+                 });
+  }
+  return requests;
+ },
+ interpretResponse: function (serverResponse, bidRequest)
+ {
+  const bidResponses = [];
+  let bidObj = bidRequest.bidRequest;
+
+  if(!serverResponse || !serverResponse.body || !bidObj)
+  {
+   utils.logWarn(`No valid bids from ${spec.code} bidder!`);
+   return [];
+  }
+  let adData = serverResponse.body;
+  let cpm = adData.cpm;
+  if(!cpm)
+  {
+   return [];
+  }
+
+  let host = spec.getBidderHost(bidObj);
+  if('mediaTypes' in bidObj && 'native' in bidObj.mediaTypes)
+  {
+   const bidResponse = {
+    requestId : bidObj.bidId,
+    cpm       : cpm,
+    width     : adData.w,
+    height    : adData.h,
+    creativeId: bidObj.params.placementId,
+    currency  : 'EUR',
+    netRevenue: true,
+    ttl       : 300,
+    native    : {
+     title             : adData.title,
+     body              : adData.body,
+     cta               : adData.cta,
+     image             : {url: adData.image},
+     clickUrl          : adData.click,
+     impressionTrackers: [adData.view]
+    },
+    mediaType : NATIVE
+   };
+  }
+  else
+  {
+   let adm = '<scr' + 'ipt>window.inDapIF=false</scr' + 'ipt><scr' + 'ipt src="//' + host + SCRIPT_URL + '"></scr' + 'ipt>' + '<ins id="' + bidObj.adspiritConId + '"></ins>' + adData.adm;
+
+   const bidResponse = {
+    requestId : bidObj.bidId,
+    cpm       : cpm,
+    width     : adData.w,
+    height    : adData.h,
+    creativeId: bidObj.params.placementId,
+    currency  : 'EUR',
+    netRevenue: true,
+    ttl       : 300,
+    ad        : adm,
+    mediaType : BANNER
+   };
+  }
+  bidResponses.push(bidResponse);
+  return bidResponses;
+ },
+ getBidderHost    : function (bid)
+ {
+  if(bid.bidder === 'adspirit')
+  {
+   return utils.getBidIdParameter('host', bid.params);
+  }
+  if(bid.bidder === 'connectad')
+  {
+   return 'connected-by.connectad.io';
+  }
+  if(bid.bidder === 'xapadsmedia')
+  {
+   return 'dsp.xapads.com';
+  }
+  return null;
+ },

```

Comment:
Are you missing a case for twiago here?

Reply:
Hello Mmoschovas,

We can also add a case for Twiago, will this result in a new audit case for you?
Our customers usually know their ad serving urls that need to be entered.

In the example given, “a.twiago.com” would have to be entered in the host and “7877” as the placement ID.

var adUnits = [
           {
               code: 'display-div',
               sizes: [[300, 250]],  // a display size
               bids: [
                   {
                       bidder: "adspirit",
                       params: {
                           placementId: '7877',
                           host: 'a.twiago.com'
                       }
                   }
               ]
           }
       ];